pkg/networkmanager/firewall-client.es6

```diff
@@ -0,0 +1,175 @@
+    enabled: false,
+    services: {},
+    enabledServices: new Set()
+};
+
+cockpit.event_target(firewall);
+
+var firewalld_service = service.proxy('firewalld');
+var firewalld_dbus = cockpit.dbus('org.fedoraproject.FirewallD1', { superuser: 'try' });
+
+firewalld_service.addEventListener('changed', function () {
+    var installed = !!firewalld_service.exists;
+
+    if (firewall.installed == installed)
+        return;
+
+    firewall.installed = installed;
+    firewall.dispatchEvent('changed');
+});
+
+function fetchServiceInfos(services) {
+    return cockpit.all(services.map(service => {
+        if (firewall.services[service])
+            return firewall.services[service];
+
+        return firewalld_dbus.call('/org/fedoraproject/FirewallD1',
+                                   'org.fedoraproject.FirewallD1',
+                                   'getServiceSettings', [service])
+            .then(reply => {
+                var [ , name, description, ports ] = reply[0];
+
+                var info = {
+                    id: service,
+                    name: name,
+                    description: description,
+                    ports: ports.map(p => ({ port: p[0], protocol: p[1] }))
+                };
+
+                firewall.services[service] = info;
+                return info;
+            });
+    }))
+
+    /*
+     * Work around `cockpit.all()` returning results in individual arguments -
+     * that's just confusing.
+     */
+    .then(function () {
+        return Array.prototype.slice.call(arguments);
+    });
+}
+
+firewalld_dbus.addEventListener('owner', function (event, owner) {
+    firewall.enabled = !!owner;
+
+    if (!firewall.enabled) {
+        firewall.services = {};
+        firewall.enabledServices = new Set();
+        firewall.dispatchEvent('changed');
+        return;
+    }
+
+    firewalld_dbus.call('/org/fedoraproject/FirewallD1',
+                        'org.fedoraproject.FirewallD1.zone',
+                        'getServices', [''])
+        .then(reply => fetchServiceInfos(reply[0]))
+        .then(services => services.map(s => firewall.enabledServices.add(s.id)))
+        .then(() => firewall.dispatchEvent('changed'))
+        .catch(error => console.warn(error));
+});
+
+firewalld_dbus.subscribe({
+    interface: 'org.fedoraproject.FirewallD1.zone',
+    path: '/org/fedoraproject/FirewallD1',
+    member: 'ServiceAdded'
+}, (path, iface, signal, args) => {
+    var [ , service, ] = args;
+
+    fetchServiceInfos([service])
+        .then(info => {
+            firewall.enabledServices.add(info[0].id);
+            firewall.dispatchEvent('changed');
+        })
+        .catch(error => console.warn(error));
+});
+
+firewalld_dbus.subscribe({
+    interface: 'org.fedoraproject.FirewallD1.zone',
+    path: '/org/fedoraproject/FirewallD1',
+    member: 'ServiceRemoved'
+}, (path, iface, signal, args) => {
+    var [ , service, ] = args;
+
+    firewall.enabledServices.delete(service);
```

Comment:
`firewall.services` needs to contain all services, because it is uses in the "Add Services" dialog. It's used as a cache if you never open the dialog so that we don't need to fetch information on every state update.